Main/ClientApp/src/Components/SpotifyElement.tsx

```diff
@@ -5,36 +5,51 @@ import { IconButton } from "@material-ui/core";
 import { VpnKey, Queue } from "@material-ui/icons";
 import SpotifyLoginDialog from "./Dialogs/SpotifyLoginDialog";
 import { addToQueue, getTrackInfo } from "../Actions/SpotifyActions";
-import { useCookie } from "../Infrastructure/CustomReactHooks";
+import Cookies from "js-cookie";
 import SpotifyStore from "../Stores/SpotifyStore";
 import { CardHeader, Avatar } from "@material-ui/core";
+import UsersStore from "../Stores/UsersStore";
+import { SpotifyTokenCookie } from "../Model/Spotify";
 
 type SpotifyElementProps = {
   element: Element;
 };
 
 export default function SpotifyElement(props: SpotifyElementProps) {
+  const [currentUser, setCurrentUser] = useState(UsersStore.getCurrentUser());
   const [spotifyLoginDialogOpen, setSpotifyLoginDialogOpen] = useState(false);
-  let [userToken, _] = useCookie<string | null>("spotifyUserToken", null);
+  let spotifyCookie = Cookies.getJSON(
+    "spotifyUserToken"
+  ) as SpotifyTokenCookie | null;
+  let userToken =
+    currentUser?.id === spotifyCookie?.userId
+      ? spotifyCookie?.spotifyToken
+      : null;
   let trackId = getTrackId(props.element.link);
   let trackUri = "spotify:track:" + trackId;
   let [track, setTrack] = useState<SpotifyApi.SingleTrackResponse | undefined>(
     SpotifyStore.getTrack(trackId)
   );
 
   useEffect(() => {
+    const trackChangeHandler = () => {
+      setTrack(SpotifyStore.getTrack(trackId));
+    };
+    const userChangeHandler = () => {
+      setCurrentUser(UsersStore.getCurrentUser());
+    };
+    SpotifyStore.addChangeListener(trackChangeHandler);
+    UsersStore.addChangeListener(userChangeHandler);
+
     if (track === undefined && userToken !== null) {
       getTrackInfo(trackId, userToken!);
     }
-    const collectionChangeHandler = () => {
-      setTrack(SpotifyStore.getTrack(trackId));
-    };
-    SpotifyStore.addChangeListener(collectionChangeHandler);
 
     return () => {
-      SpotifyStore.removeChangeListener(collectionChangeHandler);
+      SpotifyStore.removeChangeListener(trackChangeHandler);
+      UsersStore.removeChangeListener(userChangeHandler);
     };
-  }, [props.element.link]);
+  }, [trackId, userToken, track]);
```

Comment:
I don't think using track here is a good idea. It is modified in useEffect itself and causes multiple invocations. Please check if you can avoid this

Reply:
I thought the same unless I saw warning in console that i should also include track

Reply:
But it's a bad solution. It tells you to include it because track is used in useEffect, but as I commented it is a bad idea. If using state in useEffect, either state or useState should be used. Here track and setTrack are used both which can lead to bugs and unexpected results.